rhizome/host/lib/boot_image.rb

```diff
@@ -0,0 +1,124 @@
+    ext = image_ext(url)
+    init_format = initial_format(ext)
+
+    # Use of File::EXCL provokes a crash rather than a race
+    # condition if two VMs are lazily getting their images at the
+    # same time.
+    temp_file_name = @version.nil? ? @name : "#{@name}-#{@version}"
+    temp_path = File.join(image_root, "#{temp_file_name}#{ext}.tmp")
+    curl_image(url, temp_path, ca_path)
+    verify_sha256sum(temp_path, sha256sum)
+    convert_image(temp_path, init_format)
+
+    rm_if_exists(temp_path)
+  end
+
+  def image_ext(url)
+    File.extname(URI.parse(url).path)
+  end
+
+  def initial_format(ext)
+    case ext
+    when ".qcow2", ".img"
+      "qcow2"
+    when ".vhd"
+      "vpc"
+    when ".raw"
+      "raw"
+    else
+      fail "Unsupported boot_image format: #{ext}"
+    end
+  end
+
+  def curl_image(url, temp_path, ca_path)
+    ca_arg = ca_path ? " --cacert #{ca_path.shellescape}" : ""
+    File.open(temp_path, File::RDWR | File::CREAT | File::EXCL, 0o644) do
+      r "curl -f -L10 -o #{temp_path.shellescape} #{url.shellescape}#{ca_arg}"
+    end
+  end
+
+  def verify_sha256sum(temp_path, sha256sum)
+    if !sha256sum.nil? && sha256sum != Digest::SHA256.file(temp_path).hexdigest
+      fail "Invalid SHA256 sum."
+    end
+  end
+
+  def convert_image(temp_path, initial_format)
+    if initial_format == "raw"
+      File.rename(temp_path, image_path)
+    else
+      # Images are presumed to be atomically renamed into the path,
+      # i.e. no partial images will be passed to qemu-image.
+      r "qemu-img convert -p -f #{initial_format.shellescape} -O raw #{temp_path.shellescape} #{image_path.shellescape}"
+    end
+  end
+
+  # YYY: In future all images will have explicit versions and nil won't be
+  # allowed, so this method will be removed.
+  def version_to_fetch
+    @version || case @name
+                when "ubuntu-jammy"
+                  "20240319"
+                when "almalinux-9.3"
+                  "20231113"
+                end
+  end
+
+  def get_download_url
+    version = version_to_fetch
+    urls = {
+      "ubuntu-jammy" => "https://cloud-images.ubuntu.com/releases/jammy/release-#{version}/ubuntu-22.04-server-cloudimg-#{Arch.render(x64: "amd64")}.img",
+      "almalinux-9.3" => Arch.render(x64: "x86_64", arm64: "aarch64").yield_self { "https://repo.almalinux.org/almalinux/9/cloud/#{_1}/images/AlmaLinux-9-GenericCloud-9.3-#{version}.#{_1}.qcow2" }
+    }
+    urls.fetch(@name)
+  end
+
+  def get_sha256_sum
+    hashes = {
+      ["ubuntu-jammy", "x64", "20240319"] => "304983616fcba6ee1452e9f38993d7d3b8a90e1eb65fb0054d672ce23294d812",
+      ["ubuntu-jammy", "arm64", "20240319"] => "40ea1181447b9395fa03f6f2c405482fe532a348cc46fbb876effcfbbb35336f",
+      ["almalinux-9.3", "x64", "20231113"] => "6bbd060c971fd827a544c7e5e991a7d9e44460a449d2d058a0bb1290dec5a114",
+      ["almalinux-9.3", "arm64", "20231113"] => "a064715bc755346d5a8e1a4c6b1b7abffe4de03f1b0584942d5483ed32aafd67"
```

Comment:
I'll move these & the URLs to the rhizome side in a later commit and `download_image` will pass all of these values. Since this PR is meant to not change the previous behaviour (where we don't use `download_image` for ubuntu images at all), I'm keeping the URLs and hashes in the clover side.

Reply:
After some thinking, I think this map isn't needed in this PR at all. I removed it & will add it in the next PR when we do the clover changes.